### ISSUE TITLE
HUGE memory consumption and still growing memory allocations with Lottie 3.3.0

### ISSUE THREAD

**Issue by hibernat** [airbnb/lottie-ios#1461]
- [X] The issue doesn't involve an [Unsupported Feature](https://github.com/airbnb/lottie-ios/blob/master/README.md#supported-after-effects-features)
- [X] This issue isn't related to another open issue

## This issue is a:
- [] Non-Crashing Bug (Visual or otherwise)
- [X] Crashing Bug
- [] Feature Request
- [] Regression (Something that once worked, but doesn't work anymore)

## Lottie 3.3.0 is affected
Lottie 3.2.3 is better (still huge memory consumption, however not HUGE as 3.3.0)

## Platform 
- [] MacOS
- [X] iOS

## Language
- [X] Swift
- [] Objective-C

## Expected Behavior 
No memory leaks, no extensive memory allocation

## Actual Behavior
Peak memory requirement can easily be 1.5GB RAM or even more
There are many memory allocations, can be seen in the Profiler by the Allocations tool.
The app usually crashes due to lack of memory.

**Comment by Ogerets** [comment]
Same issue here

**Comment by DanielZanchi** [comment]
same issue. Trying to use the AnimationView with SwiftUI but it's not usable.

**Comment by kennic** [comment]
I had the same issue

**Comment by movch** [comment]
I have the same issue on this version

**Comment by hellbit** [comment]
The same for me, I just use `AnimationView` as a strong property in my class.

**Comment by hellbit** [comment]
Any investigation/plans about this one?

**Comment by tinder-fangchen** [comment]
We are running to a similar issue, crashing in older devices.

Below is a profile of an empty iOS app using 2 of our animations (~8KB and ~9KB each respectively)

Using lottie-ios-3.2.3
<img width="955" alt="lottie-profile-3 2 3" src="https://user-images.githubusercontent.com/53585732/162555406-bbcd54b6-adb8-4d74-a9f4-3c7fc0612ef4.png">

Using lottie-ios-3.3.0
<img width="961" alt="lottie-profile-3 3 0" src="https://user-images.githubusercontent.com/53585732/162555408-558c74ed-976a-4bd1-9340-8e3fd47fb1d3.png">

For us luckily we were able to downgrade the animation assets via backend.

But, our immediate todo is to downgrade to 3.2.3.

**Comment by tinder-fangchen** [comment]
Also I think it might be related changes in #1436

By commenting out edits around `updateRenderScale()` in...

CompositionLayer.swift
```swift
  func updateRenderScale() {
//    contentsScale = renderScale
  }
```

ShapeContainerLayer.swift
```swift
  func updateRenderScale() {
//    contentsScale = renderScale
    renderLayers.forEach( { $0.renderScale = renderScale } )
  }
```

and ShapeRendererLayer.swift
```swift
//  override func updateRenderScale() {
//    super.updateRenderScale()
//    shapeLayer.contentsScale = renderScale
//  }
```

I was able to reduce the memory from `~200 MB to ~60 MB`

Using latest `master` commenting out changes from #1436

<img width="1898" alt="comment-out-updateRenderScale-changes" src="https://user-images.githubusercontent.com/53585732/162556534-ab22b106-e091-423b-8510-3c2aac4758f1.png">

Using latest `master` no edits

<img width="1916" alt="latest-master" src="https://user-images.githubusercontent.com/53585732/162556614-3038e286-b67c-4617-b172-5687cc0a82b5.png">

**Comment by tinder-fangchen** [comment]
Diving into the history #1436 reverts changes from #1328
- #1436
- #1328 

From #1328 @absolute-heike comments https://github.com/airbnb/lottie-ios/pull/1328#issuecomment-805812191

> I could not distinguish regular Lottie animations on a pixel level when removing contentsScale (Which is not what I would have expected, so I am wondering whether I overlooked something 😅 ). Only TextLayers become blurry, which is why I removed screenScale everywhere except for TextLayers.

> Joes approach was to make renderScale publicly available. https://github.com/airbnb/lottie-ios/pull/1327
Which seems like a cleaner solution from a code point of view but has some downsides for usability. Most engineers probably won't know about this setting and you have to trade performance for blurry texts ¯\_(ツ)_/

I'm curious what the lottie team thinks. The image looks absolutely more crisp and accurate with reverted changes in #1436. But boy do we feel it from a performance standpoint.

What guidance would y'all provide? 
Would you suggest downsampling our json assets?
Maybe exposing `renderScale` config publicly is the way to go?

Thanks

**Comment by tinder-fangchen** [comment]
CC @JoeSzymanski @calda @absolute-heike

**Comment by absolute-heike** [comment]
@tinder-fangchen While official 3.2.3 fixed a bug that made sure that rendering performance and memory consumption drastically improved it still had a bug with layers that where scaled in After Effects itself.

I fixed that bug in a fork by refactoring the path rendering to how it was done in ObjC Lottie 2.x which fixed all scaling issues while improving the performance to levels even better than the ObjC Version of Lottie which in turn is much much faster then the upstream Swift Lottie.

If you want to go to a version that is crazy fast and crisp you can do that by pointing your Podfile to `pod 'tmg-lottie-ios', '3.2.3'`

**Unfortunately that version does not work with gradient strokes yet.** I didn't find time to refactor the gradient stroke rendering for the path rendering which is why I never opened a PR to upstream and by the time I got to it @calda had already reverted the change and started his own approach to refactoring the rendering.

@calda What is the current state of your rewrite and is there anything I can support you with? Do you think I should start another try at opening a PR that fixes the memory consumption and performance issues by mirroring how Paths where rendered in the ObjC version? We could do a test run, by mayve letting the UI regression tests run against the tmg-lottie-ios fork 🤔

**Comment by calda** [comment]
Earlier this year I was working on a new rendering engine that uses Core Animation (`CALayer`s  and `CAAnimation`s). In my testing, it has much better performance characteristics that the existing Main Thread engine. It's currently available behind a feature flag on `master`, and supports all layer types. It doesn't support all of the same features supported by the existing Main Thread engine, though, so it seems unlikely that we'd be able to make it the default / deprecate the existing engine.

**Comment by absolute-heike** [comment]
Do you think It makes sense to work on a new PR that fixes both the memory issue as well as the performance and blurry issues for the existing main thread engine? I think most of the code is done. I just need to rebase on the latest changes and figure out a way to refactor stroke rendering.

**Comment by calda** [comment]
Open to pull requests that improve the Main Thread engine, assuming the tests pass

**Comment by tinder-fangchen** [comment]
> @tinder-fangchen While official 3.2.3 fixed a bug that made sure that rendering performance and memory consumption drastically improved it still had a bug with layers that where scaled in After Effects itself.

> I fixed that bug in a fork by refactoring the path rendering to how it was done in ObjC Lottie 2.x which fixed all scaling issues while improving the performance to levels even better than the ObjC Version of Lottie which in turn is much much faster then the upstream Swift Lottie.

> If you want to go to a version that is crazy fast and crisp you can do that by pointing your Podfile to pod 'tmg-lottie-ios', '3.2.3'

Thank you for the suggestion @absolute-heike.

> Unfortunately gradient strokes do not work on that fork yet. I didn't find time to refactor the gradient stroke rendering for the path rendering which is why I never opened a PR to upstream and by the time I got to it @calda had already reverted the change and started his own approach to refactoring the rendering.

> Earlier this year I was working on a new rendering engine that uses Core Animation (CALayers and CAAnimations). In my testing, it has much better performance characteristics that the existing Main Thread engine. It's currently available behind a feature flag on master, and supports all layer types. It doesn't support all of the same features supported by the existing Main Thread engine, though, so it seems unlikely that we'd be able to make it the default / deprecate the existing engine.

Good to know. Thank you guys for the follow up. Noted regarding gradient stroke rendering.

For our case we need to revise our internal workflow between Design and Operations to profile performance impact of our lottie assets, which is challenging for our org since designers have little in tooling and monitoring to catch run-away performance drains for various devices. We're looking into internal tooling to avoid a similar scenario in a future release.

My suggestion to other engineering in this thread is to also revisit their animations and to profile them before release.

After presenting the issue to our team we revised the lottie animation we using; downscaling it considerably and we're seeing < 10% the mem consumption than the crashing animation; and it looks nearly identical.

Thanks for looking at this @absolute-heike , @calda .

**Comment by hannojg** [comment]
> Earlier this year I was working on a new rendering engine that uses Core Animation (`CALayer`s and `CAAnimation`s). In my testing, it has much better performance characteristics that the existing Main Thread engine. **It's currently available behind a feature flag on `master`**...

@calda sorry for disturbing this conversation! Can you tell us where to set this feature flag? Would love to test and see how it behaves in terms of resource consumption. Thanks in advance!

**Comment by calda** [comment]
Today we released Lottie 3.4.0 which includes a new rendering engine powered by Core Animation ([more details here](https://github.com/airbnb/lottie-ios/discussions/1627)). This new rendering engine has significantly better performance, and should resolve the performance issues folks have been encountering with the previous main thread rendering engine. Please try it out by setting the `LottieConfiguration.renderingEngine` feature flag, and feel free to report any new issues that you experience when using the new rendering engine.

**Comment by tinder-fangchen** [comment]
Awesome we'll test this out ASAP! Thanks Lottie team, @calda.
